include/wil/filesystem.h

```diff
@@ -1045,6 +1045,125 @@ namespace wil
+    // If the specified file does not exist and is a valid path to a writable location, a new file is created.
+    create_only_if_new = CREATE_NEW,
+
+    // Opens a file, always.
+    // If the specified file exists, the function succeeds and the last-error code is set to ERROR_ALREADY_EXISTS (183).
+    // If the specified file does not exist and is a valid path to a writable location, the function creates a file and
+    // the last-error code is set to zero.
+    open_or_create_if_needed = OPEN_ALWAYS,
+
+    // Opens a file and truncates it so that its size is zero bytes, only if it exists.
+    // If the specified file does not exist, the function fails and the last-error code is set to ERROR_FILE_NOT_FOUND (2).
+    // The calling process must open the file with the GENERIC_WRITE bit set as part of the dwDesiredAccess parameter.
+    overwrite_and_truncate_only_if_exists = TRUNCATE_EXISTING
+};
+
+/** Non-throwing open existing, test GetLastError() with the handle is invalid.
+~~~
+auto handle = wil::create_file_try_open(filePath.c_str());
+~~~
+*/
+inline wil::unique_hfile create_file_try_open(PCWSTR path, DWORD dwDesiredAccess = FILE_READ_ACCESS,
+    DWORD dwShareMode = FILE_SHARE_READ | FILE_SHARE_WRITE | FILE_SHARE_DELETE,
+    bool inheritHandle = false,
+    DWORD dwFlagsAndAttributes = FILE_ATTRIBUTE_NORMAL) noexcept
+{
+    SECURITY_ATTRIBUTES secAttributes{sizeof(secAttributes)};
+    secAttributes.bInheritHandle = inheritHandle;
+    return wil::unique_hfile{CreateFileW(
+        path, dwDesiredAccess, dwShareMode, &secAttributes, OPEN_EXISTING, dwFlagsAndAttributes, nullptr)};
+}
+
+/** open existing, throws on error.
+~~~
+auto handle = wil::create_file_open(filePath.c_str());
+~~~
+*/
+inline wil::unique_hfile create_file_open(PCWSTR path, DWORD dwDesiredAccess = FILE_READ_ACCESS,
+    DWORD dwShareMode = FILE_SHARE_READ | FILE_SHARE_WRITE | FILE_SHARE_DELETE,
+    bool inheritHandle = false,
+    DWORD dwFlagsAndAttributes = FILE_ATTRIBUTE_NORMAL)
+{
+    auto handleResult = create_file_try_open(path, dwDesiredAccess, dwShareMode, inheritHandle, dwFlagsAndAttributes);
+    THROW_LAST_ERROR_IF(!handleResult.is_valid());
+    return handleResult;
+}
+
+/** create, non-throwing, returns handle and error code as a pair.
+~~~
+auto [handle, error] = wil::create_file_try_create(filePath.c_str());
+~~~
+*/
+template<create_file_create_behavior CreateDisposition = create_file_create_behavior::overwrite_existing>
+std::pair<wil::unique_hfile, DWORD> create_file_try_create(PCWSTR path, DWORD dwDesiredAccess = FILE_READ_ACCESS | FILE_WRITE_ACCESS,
+    DWORD dwShareMode = FILE_SHARE_READ | FILE_SHARE_WRITE | FILE_SHARE_DELETE,
+    LPSECURITY_ATTRIBUTES lpSecurityAttributes = nullptr,
+    DWORD dwFlagsAndAttributes = FILE_ATTRIBUTE_NORMAL,
+    HANDLE hTemplateFile = nullptr) noexcept
+{
+    return std::pair<wil::unique_hfile, DWORD>{CreateFileW(

```

Comment:
Is this execution order (createfile first, getlasterror after) guaranteed?

Reply:
could be, yes. see my update!